sparkly/testing.py

```diff
@@ -214,65 +235,234 @@ def assertEqual(self,
             AssertionError, iff the two objects are not equal. See
             unittest.TestCase.assertEqual for details.
         """
-        def _create_key(entry):
-            if isinstance(entry, type(mock.call)):
-                _, call_args, call_kwargs = entry
-                entry = list(call_args) + list(call_kwargs.items())
-            if isinstance(entry, collections.Iterable) and not isinstance(entry, six.string_types):
-                return entry is None, tuple(_create_key(x) for x in entry)
-            return entry is None, entry
-
-        def _nicefy(obj, recursion_depth=0):
-            if isinstance(obj, float):
-                if math.isnan(obj):
-                    # assertEqual does not recognize two nans as being
-                    # equal otherwise
-                    if equal_nan:
-                        obj = 'nan'
-                elif places is not None:
-                    if significant:
-                        obj = float('%.{}g'.format(places) % obj)
-                    else:
-                        obj = round(obj, places)
-
-            if isinstance(obj, Row):
-                obj = obj.asDict(recursive=True)
-
-            if isinstance(obj, T.DataType):
-                obj = obj.jsonValue()
-
-            if isinstance(obj, type(mock.call)):
-                try:
-                    _, call_args, call_kwargs = obj
-                except ValueError:  # not called from a class instance
-                    call_args, call_kwargs = obj
-
-                obj = mock.call(
-                    *[_nicefy(call_arg, recursion_depth + 1) for call_arg in call_args],
-                    **collections.OrderedDict(_nicefy(call_kwargs, recursion_depth + 1))
+        # Our approach here is to redefine the 5 container objects that
+        # this function expects to work with - floats, dataTypes, rows,
+        # dicts and lists - to introduce generic ordering and extend
+        # the meaning of equality where applicable. We can then change
+        # all such objects to our custom containers, provide new asserters
+        # for some, and then feed them to the vanilla assertEqual.
+
+        # Define our custom containers
+        def cast_to_test_friendly_container(value, ignore_order_depth):
+            if isinstance(value, float):
+                return Float(value)
+            if isinstance(value, T.DataType):
+                return DataType(value, ignore_order_depth)
+            if isinstance(value, T.Row):
+                return Row(value, ignore_order_depth)
+            if isinstance(value, dict):
+                return Dict(value, ignore_order_depth)
+            if isinstance(value, list):
+                return List(value, ignore_order_depth)
+            return value
+
+        atol = atol or 0
+        rtol = rtol or 0
+
+        @total_ordering
+        class Float(object):
```

Comment:
The definitions of these classes (especially Floats) might depend on the arguments specified in the `assertEqual` call. I think it's cleaner this way than defining them outside of the function since they are scoped inside the only namespace they are actually used in. Having the high level plan presented in the beginning of the function helps to keep focus while going through it.